e2etest/bgptests/validate.go

```diff
@@ -356,10 +356,14 @@ func validateServiceNotInRoutesForCommunity(c *frrcontainer.FRR, community strin
 func isRouteInjected(pods []*corev1.Pod, pairingFamily ipfamily.Family, routeToCheck, vrf string) (bool, string) {
 	for _, pod := range pods {
 		podExec, err := FRRProvider.FRRExecutorFor(pod.Namespace, pod.Name)
-		Expect(err).NotTo(HaveOccurred())
```

Comment:
@fedepaol AFAIU the flake happens because even though that function is called as part of an Eventually in the main testing code, that expect just make the test fail. If we have not Expect, it retries.

I am not sure if I should do more refactor, e.g. the return "bool,string" is not ideal, it could be "string,error".

wdyt?

Reply:
not sure I understood, but the failure can't be here as here we just fetch the executor (and it will fail only if it doesn't find the pods, for example).
The failure can be in any place we call .Exec , so that's where I'd try to hide error behind a few retries (if we are not able to do better).

Reply:
>The failure can be in any place we call .Exec ,

what I see is that the executor.Exec is always wrapped in a  ginkgo Eventually logic
https://github.com/metallb/metallb/blob/main/e2etest/bgptests/validate.go#L338-L353 so even if fails a retry takes place.

As well as the function of interest https://github.com/metallb/metallb/blob/main/e2etest/bgptests/bgp.go#L1173-L1188 is part of an Eventually loop, just it has Expect instead of return error.

Therefore creating a retry logic into the call.Exec https://github.com/metallb/metallb/blob/main/e2etest/pkg/executor/executor.go#L91-L103 is a valid approach but ideally we need the code inside an Eventually/Consistently to not have other assertions, and keep the retry higher.

Reply:
(I put the comment in the wrong Expect, but still that Expect could give a flake)

Reply:
"Always" is a bold statement :-). We are calling it from many places, and I don't want to check all of them, nor worry about future ones needing to be wrapped in eventuallys. 

In particular, the job linked in the issue is failing here https://github.com/metallb/metallb/blob/main/e2etest/bgptests/validate.go#L362 where no retry is in place.

The caller must trust Exec will always work unless there is a syntax error and not to worry that it might fail temporarily. Because of this, if we won't find a better solution / the root cause of the failure, I am suggesting to do some sort of retry inside of it. And the retry should not leverage gomega, as we try to avoid gomega in util packages. 

But again, the retry is the last straw if we can't make kubectl exec work in any way (maybe bumping up kubectl? trying to file an issue against k8s?), or doing both so we have a reliable CI while we investigate?